modules/service/src/main/resources/db/migration/V1045_group_repair.sql

```diff
@@ -0,0 +1,185 @@
+
+-- This adds code that can find and close holes in groups, which can be caused by hard-deleting
+-- group elements (groups or observations) without renumbering subsequent elements ("closing holes").
+
+-- Update this procedure to add the offending group id as an exception detail.
+CREATE OR REPLACE PROCEDURE group_verify_indices(pid d_program_id) AS $$
+DECLARE
+  g d_group_id;
+BEGIN
+  WITH indices AS (
+    SELECT c_parent_id AS c_group_id, c_parent_index as c_index
+    FROM   t_group
+    WHERE  c_program_id = pid
+    UNION
+    SELECT c_group_id, c_group_index
+    FROM   t_observation
+    WHERE  c_program_id = pid
+  )
+  SELECT
+    c_group_id
+  INTO g
+  FROM
+    indices
+  GROUP BY
+    c_group_id
+  HAVING starts_at_zero_and_is_consecutive(array_agg(c_index order by c_index)) = false;
+  IF FOUND THEN
+    IF g IS NULL THEN
+      RAISE EXCEPTION 'Index discontinuity detected in the top-level group.';
+    ELSE
+      RAISE EXCEPTION 'Index discontinuity detected in group %.', g USING DETAIL = g;
```

Comment:
is this the change right? so we can get some info if we detect a problem?

Reply:
Yes, the detail field is used elsewhere so we know which group is messed up.